File: swipelibrary/src/main/java/kaufland/com/swipelibrary/dragengine/RightDragViewEngine.java
package kaufland.com.swipelibrary.dragengine;

import android.graphics.Rect;
import android.view.View;

import kaufland.com.swipelibrary.DragView;
import kaufland.com.swipelibrary.SurfaceView;
import kaufland.com.swipelibrary.SwipeDirectionDetector;
import kaufland.com.swipelibrary.SwipeLayout;
import kaufland.com.swipelibrary.SwipeResult;
import kaufland.com.swipelibrary.SwipeState;
import kaufland.com.swipelibrary.SwipeViewLayouter;

import static kaufland.com.swipelibrary.SwipeDirectionDetector.SWIPE_DIRECTION_RIGHT;
import static kaufland.com.swipelibrary.SwipeState.DragViewState.CLOSED;

/**
 * Created by sbra0902 on 29.03.17.
 */

public class RightDragViewEngine implements DraggingEngine {

    private final SwipeViewLayouter mLayouter;
    private DragView mDragView;

    private SurfaceView mSurfaceView;

    private int mInitialXPos;

    private int mDragDistance;

    private int mIntermmediateDistance;

    public RightDragViewEngine(SwipeViewLayouter layouter) {
        mLayouter = layouter;
    }

    @Override
    public void moveView(float offset, SurfaceView view, View changedView) {

        if (!mDragView.equals(changedView)) {
            mDragView.setX(view.getX() + mDragView.getWidth());
        }
    }

    @Override
    public void initializePosition(SwipeViewLayouter.DragDirection orientation) {

        mDragView = (DragView) mLayouter.getViews().get(SwipeLayout.RIGHT_DRAG_VIEW);
        mSurfaceView = mLayouter.getSurfaceView();
        mDragDistance = mDragView.getWidth();
        mInitialXPos = (int) (mSurfaceView.getX() + mDragView.getWidth());
        mIntermmediateDistance = mDragView.getSettlePointResourceId() != -1 ? mDragView.findViewById(mDragView.getSettlePointResourceId()).getRight() : mDragView.getWidth();

        moveToInitial();
    }

    private void moveToInitial() {
        mDragView.setX(mInitialXPos);
    }

    @Override
    public int clampViewPositionHorizontal(View child, int left) {
        if (left < (mSurfaceView.getWidth() - getDragDistance()) && !mDragView.isBouncePossible()) {
            return mSurfaceView.getWidth() - getDragDistance();
        }

        return left;
    }

    @Override
    public void restoreState(SwipeState.DragViewState state, SurfaceView view) {
        switch (state) {
            case LEFT_OPEN:
                mDragView.setX(mLayouter.getLeftDragView().getWidth() + mSurfaceView.getWidth());
                break;
            case RIGHT_OPEN:
                mDragView.setX(0);
                break;
            case TOP_OPEN:
                //TODO Implementation
                break;
            case BOTTOM_OPEN:
                //TODO Implementation
                break;
            default:

                mDragView.setX(mSurfaceView.getWidth());
                break;
        }
    }


    @Override
    public int getDragDistance() {
        return mDragDistance;
    }

    @Override
    public int getIntermmediateDistance() {
        return mIntermmediateDistance;
    }

    @Override
    public SwipeResult determineSwipeHorizontalState(float velocity, SwipeDirectionDetector swipeDirectionDetector, SwipeState swipeState, final SwipeLayout.SwipeListener swipeListener, View releasedChild) {
        if (releasedChild.equals(getDragView()) && swipeDirectionDetector.getSwipeDirection() == SWIPE_DIRECTION_RIGHT) {
            swipeState.setState(SwipeState.DragViewState.CLOSED);
            return new SwipeResult(mSurfaceView.getWidth(), new Runnable() {
                @Override
                public void run() {
                    swipeListener.onSwipeClosed(CLOSED);
                }
            });
        }

        return null;
    }

    @Override
    public DragView getDragView() {
        return mDragView;
    }

    @Override
    public int getOpenOffset() {
        return -mDragDistance;
    }
}

File: swipelibrary/src/main/java/kaufland/com/swipelibrary/dragengine/RightDragViewEngine.kt
package kaufland.com.swipelibrary.dragengine

import android.view.View

import kaufland.com.swipelibrary.DragView
import kaufland.com.swipelibrary.SurfaceView
import kaufland.com.swipelibrary.SwipeDirectionDetector
import kaufland.com.swipelibrary.SwipeDirectionDetector.Companion.SWIPE_DIRECTION_RIGHT
import kaufland.com.swipelibrary.SwipeLayout
import kaufland.com.swipelibrary.SwipeResult
import kaufland.com.swipelibrary.SwipeState
import kaufland.com.swipelibrary.SwipeViewLayouter

import kaufland.com.swipelibrary.SwipeState.DragViewState.CLOSED
import kaufland.com.swipelibrary.SwipeState.DragViewState.RIGHT_OPEN

/**
 * Created by sbra0902 on 29.03.17.
 */

class RightDragViewEngine(private val mLayouter: SwipeViewLayouter) : DraggingEngine<DragView> {
    override var dragView: DragView? = null

    private var mSurfaceView: SurfaceView? = null

    private var mInitialXPos: Int = 0

    override var dragDistance: Int = 0

    override var intermmediateDistance: Int = 0

    override val openOffset: Int
        get() = -dragDistance

    override fun moveView(offset: Float, view: SurfaceView, changedView: View) {

        if (dragView != changedView) {
            dragView!!.x = view.x + mSurfaceView!!.width
        }
    }

    override fun initializePosition(orientation: SwipeViewLayouter.DragDirection) {

        dragView = mLayouter.views[SwipeLayout.RIGHT_DRAG_VIEW] as DragView
        mSurfaceView = mLayouter.surfaceView
        dragDistance = dragView!!.width
        mInitialXPos = (mSurfaceView!!.x + mSurfaceView!!.width).toInt()
        intermmediateDistance = if (dragView!!.settlePointResourceId != -1) dragView!!.findViewById(dragView!!.settlePointResourceId).right else dragView!!.width

        moveToInitial()
    }

    private fun moveToInitial() {
        dragView!!.x = mInitialXPos.toFloat()
    }

    override fun clampViewPositionHorizontal(child: View, left: Int): Int {
        return if (left < mSurfaceView!!.width - dragDistance && !dragView!!.isBouncePossible) {
            mSurfaceView!!.width - dragDistance
        } else left

    }

    override fun restoreState(state: SwipeState.DragViewState, view: SurfaceView) {
        when (state) {
            SwipeState.DragViewState.LEFT_OPEN -> dragView!!.x = (mLayouter.leftDragView.width + mSurfaceView!!.width).toFloat()
            RIGHT_OPEN -> dragView!!.x = 0f
            SwipeState.DragViewState.TOP_OPEN -> {
            }
            SwipeState.DragViewState.BOTTOM_OPEN -> {
            }
            else ->

                dragView!!.x = mSurfaceView!!.width.toFloat()
        }//TODO Implementation
        //TODO Implementation
    }

    override fun determineSwipeHorizontalState(velocity: Float, swipeDirectionDetector: SwipeDirectionDetector, swipeState: SwipeState, swipeListener: SwipeLayout.SwipeListener, releasedChild: View): SwipeResult? {
        if (releasedChild == dragView && swipeDirectionDetector.swipeDirection == SWIPE_DIRECTION_RIGHT && Math.abs(swipeDirectionDetector.difX) > dragDistance / 2) {
            swipeState.state = SwipeState.DragViewState.CLOSED
            return SwipeResult(mSurfaceView!!.width, Runnable { swipeListener.onSwipeClosed(CLOSED) })
        } else if (releasedChild == dragView) {
            swipeState.state = RIGHT_OPEN
            return SwipeResult(mSurfaceView!!.width - dragDistance, Runnable { swipeListener.onSwipeClosed(RIGHT_OPEN) })
        }

        return null
    }
}
